src/AbstractPDF.php

```diff
@@ -0,0 +1,479 @@
+<?php
+
+namespace SynergiTech\ChromePDF;
+
+abstract class AbstractPDF
+{
+    /**
+     * The document format
+     *
+     * @var string
+     */
+    private $format = 'A4';
+
+    /**
+     * @var string|null
+     */
+    private $marginTop = null;
+
+    /**
+     * @var string|null
+     */
+    private $marginRight = null;
+
+    /**
+     * @var string|null
+     */
+    private $marginBottom = null;
+
+    /**
+     * @var string|null
+     */
+    private $marginLeft = null;
+
+    /**
+     * @var bool
+     */
+    private $printBackground = true;
+
+    /**
+     * @var string|null
+     */
+    private $waitUntil;
+
+    /**
+     * @var string|null
+     */
+    private $pageRanges;
+
+    /**
+     * @var string|null
+     */
+    private $emulateMedia;
+
+    /**
+     * @var float|null
+     */
+    private $scale;
+
+    /**
+     * @var bool|null
+     */
+    private $displayHeaderFooter;
+
+    /**
+     * @var string|null
+     */
+    private $header;
+
+    /**
+     * @var string|null
+     */
+    private $footer;
+
+    /**
+     * @var bool|null
+     */
+    private $preferCSSPageSize;
+
+    /**
+     * @var bool|null
+     */
+    private $landscape;
+
+    /**
+     * @var string|null
+     */
+    private $width;
+
+    /**
+    * @var string|null
+    */
+    private $height;
+
+
+    /**
+     * Sets the paper format
+     *
+     * @param  string $format
+     * @return self
+     */
+    public function setFormat(string $format): self
+    {
+        $this->format = $format;
+        return $this;
+    }
+
+    /**
+     * Sets the page margins. Supplying one argument will set all margins.
+     * Supplying two arguments will set top+bottom and right+left.
+     * Supplying all four sets them individually.
+     *
+     * @param  string|null $top
+     * @param  string|null $right
+     * @param  string|null $bottom
+     * @param  string|null $left
+     * @return self
+     */
+    public function setMargin(
+        ?string $top = null,
+        ?string $right = null,
+        ?string $bottom = null,
+        ?string $left = null
+    ): self {
+        if ($top !== null and $right === null and $bottom === null and $left === null) {
+            $this->marginTop = $top;
+            $this->marginRight = $top;
+            $this->marginBottom = $top;
+            $this->marginLeft = $top;
+        } elseif ($top !== null and $right !== null and $bottom === null and $left === null) {
+            $this->marginTop = $top;
+            $this->marginRight = $right;
+            $this->marginBottom = $top;
+            $this->marginLeft = $right;
+        } else {
```

Comment:
this function should also support three-parameter margin, i.e. top left/right bottom